### PR TITLE
Set a time interval between commands for Remotec ZXT120, and other slow devices

### PR DESCRIPTION
**Pull request by jotakar**
Remotez ZXT120 (v1.8)  result in timeout error when we send a commandSet fllowed by a commandGet
It looks like it answer an ACK before it process the command and then it doesn't act and the error comes.
Then I create a new param for nodes, named toSleep, that sets a time interval after a command is sent to the node (just before the command message is removed from queue). 
The toSleep parameter is stored in the config file zwcfg_0xxxxxx, in the <Node id="19" ...> area.
I know its working for me and another network working with Domoticz.

### CODE REVIEW

**Comment by Fishwaldo** [comment]
Can we change the function names from *Sleep to something else? Its confusing and implies to me that its something todo with wakeup/sleeping devices.

I'd prefer something like PacketDelay etc indicating its real purpose. 

Also, please use the proper DoxyGen markup in the Manager.h prototypes for documentation purposes for both functions.

**Comment by jotakar** [comment]
I've changed toSleep to PacketDelay, and I've modified comments. Now I think it's OK

**Comment by jotakar** [comment]
Is it OK?

**Comment by jotakar** [comment]
Must I make any change in this pull request or is it OK?

**Comment by jotakar** [comment]
Status of this request?

**Comment by jotakar** [comment]
I have a pull request for Domoticz that depends on this commit.

**Comment by jotakar** [comment]
Ping

File: cpp/src/Node.cpp [hunk 1]
```diff
@@ -1836,6 +1847,22 @@ void Node::SetLocation
...
+// Set PacketDelay value for the node
+//-----------------------------------------------------------------------------
+void Node::SetPacketDelay
+(
+        uint32 const& _PacketDelay
+)
+{
+    	m_PacketDelay = _PacketDelay;
+    	// Notify the watchers of the value changes
+    	Notification* notification = new Notification( Notification::Type_NodeNaming );
```
Comment (Fishwaldo) [inline cpp/src/Node.cpp]: Can you drop this notification? Its not needed.

**Comment by jotakar** [comment]
Oh!, Hi: glad to read your message!. 
I'll try to explain this code:
I'm using OZWcontrol Panel to change the value of PacketDelay ( It's an UI used in Domoticz for acces to devices status, properties and configuration).
If I drop this notification then, when I change the value of PacketDelay the new value is not reflected inmediately in control panel. The easiest way I found to notify OZWControl panel was using a NodeNaming notification, like does node name or location changes.  
Now, whit this notification,  when I change the PacketDelay value the OZWControl panel shows the new value inmediately, This is the same behaviour than when I change name or location.
This is the reason because of I think this code is needed.

Thanks.

**Comment by Fishwaldo** [comment]
The Notifications like NodeNaming are more associated with the progress of querying them when bringing them online. And NodeNaming would typically only happen once per restart of the application. 

What your describing is more application specific, and your abusing this notification. It's possible it might break other applications that use OZW as well. 

Either you should fix the application (this is a immediate change, so no need for a callback to confirm the change has been effected) or implement the change as a Hard Coded Configureation ValueID. 

I'd prefer the first option tho, as that's how other options that affect individual nodes operate.

**Comment by jotakar** [comment]
I don't undesrstand at all.
I see that NodeNaming notification occurs when we change the **name** or **location** of a device, they doesn't occur only when restart application. So I thought that PacketDelay could work like this other devices values does. I see PacketDelay as a property like Name or Location. 
That's my doubt.
But if this is not a right method, then I'll delete notification and I'll look for a workaround.
Thanks

**Comment by jotakar** [comment]
Heloo:
Changes commited.
Thanks

**Comment by jotakar** [comment]
I made the change :)

**Comment by jotakar** [comment]
Ping

**Comment by jotakar** [comment]
Needed more changes?

**Comment by jotakar** [comment]
Ping.
Is it OK now?

**Comment by jotakar** [comment]
Ping

**Comment by jotakar** [comment]
Is it necessary another change? Last required change was done some time ago.
I need this commit to do a related pull request in Domoticz, 
Thanks

**Comment by jotakar** [comment]
Which is the problem now? Who can commit this pull request?
Thanks

**Comment by jotakar** [comment]
Ping, ping

**Comment by jotakar** [comment]
Please, any news about this ull request?
Thanks

**Comment by jotakar** [comment]
On Feb 17, I made the change requested by @Fishwaldo , there is no conflict with the bas branch and the modification is working fine. When this modification is incorporated to OZwave I have to made another pull request for the OZW panel control. 
Thanks

**Comment by jotakar** [comment]
Why is not commited  this pull request?
Thanks

**Comment by jotakar** [comment]
Please, what is happening with this request?
Thanks

**Comment by jotakar** [comment]
Ping

**Comment by jotakar** [comment]
This pull request is three months old. Last review is one ad a half months old.
Any news?
Thanks

**Comment by jotakar** [comment]
Is there anyone there?

**Comment by jotakar** [comment]
More than one month from last update, now I wonder which is the status of this pull request.  ?????

**Comment by jotakar** [comment]
Must I do any other change or anything else to get this pull request accepted and merged?
Thanks

**Comment by nechry** [comment]
@jotakar I'm sorry, but when @Fishwaldo will have time, it will review and merge your PR. Until then, stop asking every 2 days for an update. We do this activity on free time, and we set priorities.

**Comment by jotakar** [comment]
Ok, I’m sorry. 

I was not sure if I had to do anything else or so, I have sent very few pull request to Github

Thanks for your email, and again I beg your pardon.

 

 

 

 

De: Jean-François Auger [mailto:notifications@github.com] 
Enviado el: martes, 25 de abril de 2017 14:23
Para: OpenZWave/open-zwave
CC: jotakar; Mention
Asunto: Re: [OpenZWave/open-zwave] Set a time interval between commands for Remotec ZXT120, and other slow devices (#1090)

 

@jotakar <https://github.com/jotakar>  I'm sorry, but when @Fishwaldo <https://github.com/Fishwaldo>  will have time, it will review and merge your PR. Until then, stop asking every 2 days for an update. We do this activity on free time, and we set priorities.

File: cpp/src/Node.cpp (hunk 1 above)
Comment (jotakar) [inline cpp/src/Node.cpp]: Hi, I made this change on Feb 17 (the last change requested), since then I'm waiting for an answer,
Thanks

**Comment by iw3row** [comment]
Hi @Fishwaldo  I have the exact same issue with Remotec ZXT-120, is there any chance the solution proposed by @jotakar will be merged to the master soon? Many thanks in advance for your help and time dedicated to the project!

**Comment by cscattaglia917** [comment]
@jotakar - thanks for your changes - I am eager to try this out and hopefully get my ZXT-120 reporting values back to me.
One question - how do I set the 'toSleep/packetDelay' parameter in the zwcfg file?  

Thanks!

**Comment by jotakar** [comment]
@cscattaglia917 you have to set this value by hand. It is in zwcfg_homeid.xml file (I'm working with a Linux based system). homeid is the identifier of your net, it's in Config folder. The line looks like
`    <Node id="19" name="ACSalon" location="Salon" basic="4" generic="8" specific="6" type="General Thermostat V2" listening="true" frequentListening="false" beaming="true" routing="true" max_baud_rate="40000" version="4" PacketDelay="5000" query_stage="Complete">`
If this modification  is commited, I'll send a modification for OZwave Control Panel to change this value from this application.

File: cpp/src/Node.cpp (hunk 1 above)
Comment (jotakar) [inline cpp/src/Node.cpp]: Why this request is waiting for being commited? 
Thanks

**Comment by cscattaglia917** [comment]
Thanks @jotakar - that helps a lot.  Although, I am using HomeAssistant and my zwcfg_*.xml file is overwritten upon startup of the application, so my manual changes are removed.. Is there somewhere else I should be modifying this at so that it is written to the zwcfg file on startup?

Thanks,

**Comment by jotakar** [comment]
I don't know HomeAssistant (I'm using domoticz), so I wonder if you can set a name or location to your devices. If so, does these data changes upon start up the aplilcation? 
I've read in HomeAssistant home page that theres is a configuration file named configuration.yaml, perhaps you can try this file.
And in the HA webpage I've read also something about ZWave... 
Thats all I can tell you, I don't know Home Asistant. 
Bye

**Comment by cscattaglia917** [comment]
I was able to get my changes to stay by making the file un-writable by the HomeAssistant user. (Definitely not a solution, but was just testing.) 
There is a method to rename from within HA UI, and that change does remain in the zwcfg config file after a restart, but I can't add custom values. 

Even with forcing the Packet delay parameter, it still doesn't appear to work for me. The thermostat receives the temperature I set, but it never changes, and OZW returns the current temperature of the sensor as the 'setpoint' temperature. 
I think it's on me to figure it out from here though..

**Comment by Dpons039** [comment]
Hey there, I'm another user with the same issues with this Device. It would be awesome to have the pull finally merged!

Thank you all for your effords!

File: cpp/src/Node.cpp (hunk 1 above)
Comment (balintkozma) [inline cpp/src/Node.cpp]: Merge it already!

**Comment by balintkozma** [comment]
I also had to chown the zwcfg to root, to make PacketDelay persistent. This solution is not ready.

- put the PacketDelay into the xsd files: Node in zwcfg.xsd, Protocol in device_configuration.xsd, 
- wire them together in Node.cpp, like nodeinfosupported
- put a Protocol element with default PacketDelay value into the zxt-120.xml. 
- I experienced timeouts with 5000 usec, I'm going with 50000 now.

**Comment by jotakar** [comment]
Ping

File: cpp/src/Node.cpp (hunk 1 above)
Comment (jotakar) [inline cpp/src/Node.cpp]: Is it necessary any other modification?
Please, could you answer?

**Comment by ptje** [comment]
Hello,

I have the same problem, when can the modification get true ? It resolves lot of problems for several users ?

Greetings,
Peter

**Comment by balintkozma** [comment]
@jotakar: what about putting PacketDelay into zxt-120.xml? See my earlier comment.

**Comment by jotakar** [comment]
@balintkozma: After this modification is incorprated to OZW library we could open a new pull request to modify ZXT config files. 
I had another pull request for Domoticz to manage PacketDelay, but I cannot send it because I have to wait for @Fishwaldo commiting this modification.... and this is the problem... why does @Fishwaldo froze it? I don't know.

**Comment by balintkozma** [comment]
Seriously, what is happening here? Did this somewhere get personal or what?

I had a v1.7, later got a v1.8 version of ZXT-120, and they were both unusable with UZB1 stick. This issue is not an enhancement, rather a major defect. 

@Fishwaldo: please, help this thread
@jotakar: this setting really should be in the xml, as an optional protocol setting, in this pull request.

**Comment by jotakar** [comment]
I can not understand why this pull is frozen for so long (more than 5 months!).
It takes less than five minutes to see that all requested changes have been done and commit it. It is not a long time, I think.

**Comment by tbrasser** [comment]
For what it's worth. I'm also waiting for this pull-request to be merged. (Or actually, for the zxt-120 to work reliable with OpenZ-Wave) before purchasing 2 of them. I also have not found another solution to bring Air-Conditioning units into my home automation.

**Comment by jotakar** [comment]
Hi
this enhancement is necessary for people using ZXT120, and it's waiting for being merged more than SIX MONTHS, and I can't understand the reason. If @Fishwaldo has no time to attend requests for ZWave library he could pass the work to another reviewer. Please, my work is done, could @Fishwaldo  do his work? Several users is are waiting for this modification. 
Thanks

**Comment by jurgenweber** [comment]
Can we please get this resolved and merged please.

**Comment by dcarrion87** [comment]
Does anyone know if this is getting merged soon?

**Comment by jotakar** [comment]
This fix is working but Fishwaldo doesn't like it. Read the pull request #1310. 
You can apply this modification, it works for me. 
And you can follow the pull request #1315,  a more generic version of #1310 . If it's merged I'l try to find a solution using this Timer Thread enhancement. But I can't promise  result.

**Comment by AriZuu** [comment]
It would be great to get ZXT120 working, by this PR or based on another ones. I just purchased one and after trying to get it working ended up here waiting for fix :-(

I can help with coding & testing if needed.

**Comment by turbokongen** [comment]
I have three of these, and they all work,(2pcs v1.6 1pcs v1.8) but with small hickups. It became way better, and the device reported back it's setpoint when I modified: `getSupported=true` in the config file.
The hickups are that sometimes the device sends the same command to the AC multiple times. But I don't mind that as long as it works.

**Comment by AriZuu** [comment]
@turbokongen tried with getsupported=true, but it doesn't really work much better. Mine is v1.8. Sometimes setting temperature or mode goes through, but mostly not.

**Comment by AriZuu** [comment]
Idea in #1315 sounds rather elegant, so using it as base for this fix might be a good idea.

**Comment by dcarrion87** [comment]
Confirming getsupported doesn't really make significant difference.

**Comment by aosadchyy** [comment]
I got the same issue with newer Remotec ZTS-500. tweaking write_only helps to make set work. Here is the post
https://github.com/OpenZWave/open-zwave/issues/1063
@Fishwaldo 
Why is this pull request not completed yet? Lots of people are suffering from this issue.

**Comment by turbokongen** [comment]
I gave this a shot, and here is my findings:
The timeout works good for making the ZXT-120 respond like it should. However, if I try to include a new node with this patch running, it get's stuck.
```
2017-12-22 16:03:54.703 Info, Node054, Received reply to FUNC_ID_ZW_GET_NODE_PROTOCOL_INFO
2017-12-22 16:03:54.703 Info, Node054,   Protocol Info for Node 54:
2017-12-22 16:03:54.703 Info, Node054,     Listening     = false
2017-12-22 16:03:54.703 Info, Node054,     Frequent      = false
2017-12-22 16:03:54.703 Info, Node054,     Beaming       = true
2017-12-22 16:03:54.703 Info, Node054,     Routing       = true
2017-12-22 16:03:54.703 Info, Node054,     Max Baud Rate = 40000
2017-12-22 16:03:54.703 Info, Node054,     Version       = 4
2017-12-22 16:03:54.703 Info, Node054,     Security      = false
2017-12-22 16:03:54.709 Info, Node054,   Basic device class    (0x04) - Routing Slave
2017-12-22 16:03:54.710 Info, Node054,   Generic device Class  (0x07) - Notification Sensor
2017-12-22 16:03:54.710 Info, Node054,   Specific device class (0x01) - Notification Sensor
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_BASIC is not mapped
2017-12-22 16:03:54.710 Info, Node054,   Mandatory Command Classes for Node 54:
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_BASIC
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_SENSOR_BINARY
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_DEVICE_RESET_LOCALLY
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_ZWAVE_PLUS_INFO
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_CONFIGURATION
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_ALARM
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_MANUFACTURER_SPECIFIC
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_POWERLEVEL
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_BATTERY
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_WAKE_UP
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_ASSOCIATION
2017-12-22 16:03:54.710 Info, Node054,     COMMAND_CLASS_VERSION
2017-12-22 16:03:54.710 Info, Node054,   Mandatory Command Classes controlled by Node 54:
2017-12-22 16:03:54.710 Info, Node054,     None
2017-12-22 16:03:54.710 Detail, Node156,   Expected reply was received
2017-12-22 16:03:54.710 Detail, Node156,   Message transaction complete
2017-12-22 16:03:54.710 Detail, 
2017-12-22 16:05:56.231 Info, mgr,     Manager::WriteConfig completed for driver with home ID of 0xd5a2c6e5
2017-12-22 16:05:57.238 Info, mgr,     Driver for controller /dev/serial/by-id/usb-0658_0200-if00 pending removal
2017-12-22 16:05:57.239 Detail, Node054, Notification: NodeProtocolInfo
2017-12-22 16:05:57.239 Detail, Notification: DriverRemoved
```
The Driver info got printed because I stopped the application. Otherwise it just hang the ozw program and no send or receive.
This was with this PR rebased together with latest master.
I can't really say where the issue is in the code, but somebody else might see that. It is atleast a major defect that should be fixed before merging.

**Comment by B1tMaster** [comment]
will this ever be merged?  does this solve the issue of Remotec ZXT 120 never getting the zwave commands? 

https://community.home-assistant.io/t/remotec-zxt-120-operational-questions-my-script/20260/12

**Comment by jotakar** [comment]
This request has a year old!!!!
It looks like @Fishwaldo has not time enough to follow this request... or the #1315, a related fix that I was waiting for in order to improve this fix for Remotec ZXT 120.

**Comment by 5ke** [comment]
Just adding myself to the list of people who'd love to see this fix merged. In the mean time I'll try and manually pull this fix into home assistant - any help on how to do that will be appreciated :-)

**Comment by balintkozma** [comment]
Just add that one line: usleep(50000); 
ZXT-120 will work, and it won't brake anything.

**Comment by 5ke** [comment]
You mean in cpp/src/Driver.cpp and then without checking which node, but just applying it to all? That sounds relatively easy :-) (however, for now I'm stuck on trying to get to the openzwave files on my hassio system)

**Comment by balintkozma** [comment]
Yes. Ugly, but works, and simple.

**Comment by jotakar** [comment]
I'm confused with this pull request. It appears that there are changes requested, but all requested changes HAS BENN DONE LONG TIME AGO. 
So I wonder why this pull is open and waiting, anybody could answer?
Thanks

**Comment by jotakar** [comment]
How many years need Fishwaldo to answer this addon?

**Comment by jotakar** [comment]
ping?

**Comment by satzg** [comment]
It will be the same answer as how many years need Jotakar keep asking the question. :-)

Thanks,
Greg

**Comment by jotakar** [comment]
@Fishwaldo  please, could you see if this pull will be commited? Or at least, could you give an answer about the status of this request? More than 2 years!!!! I think it's too long time, dont you think so?

**Comment by mrBaas** [comment]
@5ke did you manage to find the openzwave files in hassio? 
If so did you manage to get it working with the usleep(5000) change..?

I need this to work asap.

**Comment by 5ke** [comment]
@mrBaas: see my post on the home assistant forum: https://community.home-assistant.io/t/applying-fix-in-openzwave-to-get-remotec-zxt-120-working-with-hassbian/49048 

I gave up on trying to get it done in Hassio. Managed to get into the docker, but got the feeling I didn't have the right rights to make the necessary changes. Switched to Hassbian and got as far as the forum post shows... Let me know how you get on :-)

**Comment by jotakar** [comment]
All requested for this requeste changes were made long time ago. But  it looks like @Fishwaldo has no time to review them and commint or close or answer at least. It' very easy, just two minutes for writing a post explaining which is the situation of this request. Any more.
I wonder if he is still a collaborator of this project

**Comment by JonLaliberte** [comment]
@nechry Any chance you can assist here?

**Comment by olegk0** [comment]
@jotakar
Hi , please add initialization of m_PacketDelay (it useable if config do not exist)

```
--- Node.cpp.old	2018-07-08 20:29:04.614691529 +0200
+++ Node.cpp	2018-07-08 20:20:23.048931677 +0200
@@ -152,6 +152,7 @@
 m_productName( "" ),
 m_nodeName( "" ),
 m_location( "" ),
+m_PacketDelay ( 0 ),
 m_manufacturerId( 0 ),
 m_productType( 0 ),
 m_productId( 0 ),
```

**Comment by jotakar** [comment]
Thanks @olegk0  for your suggestion. This is a good idea. I modified Node.cpp, but I think it's useless because It looks like @Fishwaldo  doesn't care about this pull request. :(

**Comment by olegk0** [comment]
Hi @jotakar , to be honest, I'm also not sure the delay method is optimal (although it works!).
Perhaps this problem is in poor compatibility of zwave API versions.

**Comment by jotakar** [comment]
Perhaps it's not a beautiful code, but it works and, I'm sincere, I worked a lot studing how ZWave libs works and this is the only code I was able to get ... 
And, if there is a compatibility problem or any other one, I cannot understand why @Fishwaldo doesn't tell anything about it!.
I would like to learn more about zwave API in order to get a more compatible and beautiful solution, but meanwhile this is the only solution I've found. 
Is there any other way to get a particular behavior for a particular device with a particular problem? I would like to find any other method, indeed I would.
Thanks again for your aid.

**Comment by Fishwaldo** [comment]
Closing.

**Pull request closed without merging**